In [2]:
# Загружаем нужные библиотеки

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

### Датасет был взят из Kaggle
#### New York City Airbnb Open Data. Airbnb listings and metrics in NYC, NY, USA (2019)
Этот файл данных содержит всю необходимую информацию, чтобы узнать больше о хостингах, географической доступности, необходимых показателях для составления прогнозов и выводов.)

In [3]:
# Загружаем датасет

df = pd.read_csv('AB_NYC_2019.csv')
df.drop_duplicates(inplace=True)
df.head(5)

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

### О чем датасет
*Airbnb — онлайн-площадка для размещения и поиска краткосрочной аренды частного жилья по всему миру.*

1. name - Название листинга
2. host_id - Id хозяина
3. host_name - Имя хозяина 
4. neighbourhood_group - локация(Манхэттен, Бруклин, ....)
5. neighbourhood - Регион(Харлем, ....)
6. latitude, longitude - Долгота и широта
7. room_type - Тип жилья(Целый дом, приватная комната, делишь комнату)
8. price - Цена за один день
9. minimum_nights - Количество минимальных дней которых можно жить
10. number_of_reviews - Количество отзывов
11. last_review - Дата последнего отзыва
12. reviews_per_month - Количество отзывов за месяц
13. calculated_host_listings_count - количество объявлений на один хостинг
14. availability_365 - количество дней, в течение которых список доступен для бронирования




In [4]:
# Немного инфы о датасете

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48895 entries, 0 to 48894
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   id                              48895 non-null  int64  
 1   name                            48879 non-null  object 
 2   host_id                         48895 non-null  int64  
 3   host_name                       48874 non-null  object 
 4   neighbourhood_group             48895 non-null  object 
 5   neighbourhood                   48895 non-null  object 
 6   latitude                        48895 non-null  float64
 7   longitude                       48895 non-null  float64
 8   room_type                       48895 non-null  object 
 9   price                           48895 non-null  int64  
 10  minimum_nights                  48895 non-null  int64  
 11  number_of_reviews               48895 non-null  int64  
 12  last_review                     

In [5]:
# Сумма пропущенных значений

df.isnull().sum()

id                                    0
name                                 16
host_id                               0
host_name                            21
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10052
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

In [6]:
# Смотрим уникальные значения

print("Список всех локации:")
print(df['neighbourhood_group'].unique())

print()
print("Список всех типов жилья:")
print(df['room_type'].unique())

print()
print("Список всех регионов:")
print(df['neighbourhood'].unique())



Список всех локации:
['Brooklyn' 'Manhattan' 'Queens' 'Staten Island' 'Bronx']

Список всех типов жилья:
['Private room' 'Entire home/apt' 'Shared room']

Список всех регионов:
['Kensington' 'Midtown' 'Harlem' 'Clinton Hill' 'East Harlem'
 'Murray Hill' 'Bedford-Stuyvesant' "Hell's Kitchen" 'Upper West Side'
 'Chinatown' 'South Slope' 'West Village' 'Williamsburg' 'Fort Greene'
 'Chelsea' 'Crown Heights' 'Park Slope' 'Windsor Terrace' 'Inwood'
 'East Village' 'Greenpoint' 'Bushwick' 'Flatbush' 'Lower East Side'
 'Prospect-Lefferts Gardens' 'Long Island City' 'Kips Bay' 'SoHo'
 'Upper East Side' 'Prospect Heights' 'Washington Heights' 'Woodside'
 'Brooklyn Heights' 'Carroll Gardens' 'Gowanus' 'Flatlands' 'Cobble Hill'
 'Flushing' 'Boerum Hill' 'Sunnyside' 'DUMBO' 'St. George' 'Highbridge'
 'Financial District' 'Ridgewood' 'Morningside Heights' 'Jamaica'
 'Middle Village' 'NoHo' 'Ditmars Steinway' 'Flatiron District'
 'Roosevelt Island' 'Greenwich Village' 'Little Italy' 'East Flatbush'


In [7]:
# Немного статистики по нужным столбцам 

df[["price", "minimum_nights", "number_of_reviews", "reviews_per_month", "calculated_host_listings_count", "availability_365"]].describe()

price  minimum_nights  number_of_reviews  reviews_per_month  \
count  48895.000000    48895.000000       48895.000000       38843.000000   
mean     152.720687        7.029962          23.274466           1.373221   
std      240.154170       20.510550          44.550582           1.680442   
min        0.000000        1.000000           0.000000           0.010000   
25%       69.000000        1.000000           1.000000           0.190000   
50%      106.000000        3.000000           5.000000           0.720000   
75%      175.000000        5.000000          24.000000           2.020000   
max    10000.000000     1250.000000         629.000000          58.500000   

       calculated_host_listings_count  availability_365  
count                    48895.000000      48895.000000  
mean                         7.143982        112.781327  
std                         32.952519        131.622289  
min                          1.000000          0.000000  
25%                          1.000000          0.000000  
50%                          1.000000         45.000000  
75%                          2.000000        227.000000  
max                        327.000000        365.000000

### Some outliers
1. Максимальная цена 10000 долларов за ночь и минимальная цена равна нулю, подозрительно нужно будет проверить
2. Максимальное значение столбца minimum_nights 1250 что просто невозможно, нужно будет проверить 
3. Максимальное значение calculated_host_listings_count 327, тоже  подозрительно при среднем 7 а медианы 1

### Some statistics
1. Средняя цена завышена по сравнению с медианой скорее из за выбросов, и стандартное отклонение 240 что тоже доказывает это. 
2. Среднее менее чувствительна к выбросам и более четко показывает "среднюю цену"
3. По Q1 и Q3 видно что 50% цен лежит в диапазоне 69 и 175 долларов
4. Если посчитать нижнюю и верхнюю границу по формулам Q1 - 1.5 * IQR и Q3 + 1.5 * IQR (IQR = Q3 − Q1) то нижняя граница равна -90 но цены не могут быть минусовыми так что 0 и верхняя 334.
Так что, что не входит в эти границы скорее считаются выбросами но нужно будет сделать подробный анализ.

In [8]:
# Проверка на средную и медиану по типу жилья 

print("Количество значений по типу жилья:")
print(df["room_type"].value_counts())

print()
print('Медиана по типу жилья:')
print("Entire home/apt:", df[df['room_type'] == "Entire home/apt"]['price'].median())
print("   Private room:", df[df['room_type'] == "Private room"]['price'].median())
print("    Shared room:", df[df['room_type'] == "Shared room"]['price'].median())

print() 
print('Среднее по типу жилья:')
print("Entire home/apt:", df[df['room_type'] == "Entire home/apt"]['price'].mean())
print("   Private room:", df[df['room_type'] == "Private room"]['price'].mean())
print("    Shared room:", df[df['room_type'] == "Shared room"]['price'].mean())

Количество значений по типу жилья:
room_type
Entire home/apt    25409
Private room       22326
Shared room         1160
Name: count, dtype: int64

Медиана по типу жилья:
Entire home/apt: 160.0
   Private room: 70.0
    Shared room: 45.0

Среднее по типу жилья:
Entire home/apt: 211.79424613325986
   Private room: 89.78097285675894
    Shared room: 70.12758620689655


In [9]:
# Немного статистики с локацией, типом жилья и ценой

print("Среднее по локациям:")
print(df.groupby('neighbourhood_group')['price'].mean())
print()

print("Медиана по локациям:")
print(df.groupby('neighbourhood_group')['price'].median())
print()

print("Среднее по локациям и по типу жилья:")
print(df.groupby(['neighbourhood_group', 'room_type'])['price'].mean())

print()
print("Медиана по локациям и по типу жилья:")
print(df.groupby(['neighbourhood_group', 'room_type'])['price'].mean())

Среднее по локациям:
neighbourhood_group
Bronx             87.496792
Brooklyn         124.383207
Manhattan        196.875814
Queens            99.517649
Staten Island    114.812332
Name: price, dtype: float64

Медиана по локациям:
neighbourhood_group
Bronx             65.0
Brooklyn          90.0
Manhattan        150.0
Queens            75.0
Staten Island     75.0
Name: price, dtype: float64

Среднее по локациям и по типу жилья:
neighbourhood_group  room_type      
Bronx                Entire home/apt    127.506596
                     Private room        66.788344
                     Shared room         59.800000
Brooklyn             Entire home/apt    178.327545
                     Private room        76.500099
                     Shared room         50.527845
Manhattan            Entire home/apt    249.239109
                     Private room       116.776622
                     Shared room         88.977083
Queens               Entire home/apt    147.050573
                     

#### У нас есть значение None в name и host_name
#### Посчитал что это это не прям критично заменю их на 'unspecified'

In [10]:
# 

df.name.fillna('unspecified', inplace=True)
df.host_name.fillna('unspecified', inplace=True)

df.isnull().sum()

id                                    0
name                                  0
host_id                               0
host_name                             0
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10052
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

In [11]:
df[df["last_review"].isnull() == True][:5]

id                                               name  host_id  \
2    3647                THE VILLAGE OF HARLEM....NEW YORK !     4632   
19   7750                  Huge 2 BR Upper East  Cental Park    17985   
26   8700  Magnifique Suite au N de Manhattan - vue Cloitres    26394   
36  11452                        Clean and Quiet in Brooklyn     7355   
38  11943                          Country space in the city    45445   

          host_name neighbourhood_group       neighbourhood  latitude  \
2         Elisabeth           Manhattan              Harlem  40.80902   
19             Sing           Manhattan         East Harlem  40.79685   
26  Claude & Sophie           Manhattan              Inwood  40.86754   
36               Vt            Brooklyn  Bedford-Stuyvesant  40.68876   
38          Harriet            Brooklyn            Flatbush  40.63702   

    longitude        room_type  price  minimum_nights  number_of_reviews  \
2   -73.94190     Private room    150               3                  0   
19  -73.94872  Entire home/apt    190               7                  0   
26  -73.92639     Private room     80               4                  0   
36  -73.94312     Private room     35              60                  0   
38  -73.96327     Private room    150               1                  0   

   last_review  reviews_per_month  calculated_host_listings_count  \
2          NaN                NaN                               1   
19         NaN                NaN                               2   
26         NaN                NaN                               1   
36         NaN                NaN                               1   
38         NaN                NaN                               1   

    availability_365  
2                365  
19               249  
26                 0  
36               365  
38               365

#### Так как у нас 10000 нулл значении в review per month то здесь скорее всего просто не было вообще отзывов либо могла произойти ошибка при скрапинге.
#### Я посчитал нужным заменить их на ноль так как непонятна причина пропуска 
#### Так же и со столбцом last_review  заменю просто самой старой датой из датасета

In [20]:
df['reviews_per_month'].fillna(0, inplace=True)
df['reviews_per_month'].isnull().sum()

df['last_review'] = pd.to_datetime(df['last_review'], errors='coerce')

df['last_review'].fillna(df['last_review'].max(), inplace=True)
print(df.isnull().sum())

id                                0
name                              0
host_id                           0
host_name                         0
neighbourhood_group               0
neighbourhood                     0
latitude                          0
longitude                         0
room_type                         0
price                             0
minimum_nights                    0
number_of_reviews                 0
last_review                       0
reviews_per_month                 0
calculated_host_listings_count    0
availability_365                  0
dtype: int64


### Работа с выброосами в столбце прайс

In [35]:
shared_room = df[df['room_type'] == 'Shared room']

shared_room[shared_room['price'] >= 1000]

id                                    name   host_id host_name  \
3700    2230982       An Accomodating Apartment on Wall  11385753      Greg   
14380  11234747  Mins away to Manhattan Suite Residence  24146326    Julien   
23545  19057018             1BR/1BA in 2BR/2BA Flushing    356490     Peter   
30080  23169146                 Artist apartment in LIC  51260506     Grace   

      neighbourhood_group       neighbourhood  latitude  longitude  \
3700            Manhattan  Financial District  40.70511  -74.00943   
14380              Queens             Astoria  40.76626  -73.93054   
23545              Queens            Flushing  40.76796  -73.81732   
30080              Queens    Long Island City  40.74786  -73.94628   

         room_type  price  minimum_nights  number_of_reviews last_review  \
3700   Shared room   1000               1                  0  2019-07-08   
14380  Shared room   1800               3                  5  2017-04-09   
23545  Shared room   1000               1                  0  2019-07-08   
30080  Shared room   1250               1                  1  2018-02-19   

       reviews_per_month  calculated_host_listings_count  availability_365  
3700                0.00                               1                 0  
14380               0.13                               2                90  
23545               0.00                               1                 0  
30080               0.06                               1               173

In [37]:
private_room = df[df['room_type'] == 'Private room']

private_room[private_room['price'] >= 3000]

id                                              name    host_id  \
946      363673                  Beautiful 3 bedroom in Manhattan     256239   
4345    2919330                  NearWilliamsburg bridge 11211 BK   14908606   
9151    7003697               Furnished room in Astoria apartment   20582832   
12342   9528920               Quiet, Clean, Lit @ LES & Chinatown    3906464   
28727  22207357                                     The room boom  162116640   
28946  22296097  Chelsea Gallery Space for events and exhibitions    3750764   
37194  29547314              Apartment New York \nHell’s Kitchens   35303743   
39846  30962126                                      CQ 1 Bedroom   30223355   
43670  33796251  Beautiful private Brooklyn room with kitchenette    8748976   
45666  34895693                              Gem of east Flatbush  262534951   

      host_name neighbourhood_group       neighbourhood  latitude  longitude  \
946      Tracey           Manhattan     Upper West Side  40.80142  -73.96931   
4345     Bianca            Brooklyn  Bedford-Stuyvesant  40.69572  -73.95731   
9151   Kathrine              Queens             Astoria  40.76810  -73.91651   
12342       Amy           Manhattan     Lower East Side  40.71355  -73.98507   
28727     Tuvia            Brooklyn      Brighton Beach  40.58363  -73.96405   
28946     Kevin           Manhattan             Chelsea  40.74913  -74.00373   
37194  Patricia           Manhattan     Upper West Side  40.76835  -73.98367   
39846  Adrienne           Manhattan  Financial District  40.70706  -74.00991   
43670   Jeffrey            Brooklyn  Bedford-Stuyvesant  40.68807  -73.95426   
45666    Sandra            Brooklyn       East Flatbush  40.65724  -73.92450   

          room_type  price  minimum_nights  number_of_reviews last_review  \
946    Private room   3000               7                  0  2019-07-08   
4345   Private room   5000               6                 10  2016-01-02   
9151   Private room  10000             100                  2  2016-02-13   
12342  Private room   9999              99                  6  2016-01-01   
28727  Private room   3000               1                  0  2019-07-08   
28946  Private room   3210               1                  0  2019-07-08   
37194  Private room   6500              30                  0  2019-07-08   
39846  Private room   3000               3                  0  2019-07-08   
43670  Private room   4200             114                  0  2019-07-08   
45666  Private room   7500               1                  8  2019-07-07   

       reviews_per_month  calculated_host_listings_count  availability_365  
946                 0.00                               1               365  
4345                0.16                               1               363  
9151                0.04                               1                 0  
12342               0.14                               1                83  
28727               0.00                               1                90  
28946               0.00                               6               364  
37194               0.00                               1                97  
39846               0.00                               1                23  
43670               0.00                               1               347  
45666               6.15                               2               179

In [38]:
df[df['price'] == 0]

id                                               name    host_id  \
23161  18750597  Huge Brooklyn Brownstone Living, Close to it all.    8993084   
25433  20333471      ★Hostel Style Room | Ideal Traveling Buddies★  131697576   
25634  20523843    MARTIAL LOFT 3: REDEMPTION (upstairs, 2nd room)   15787004   
25753  20608117                    Sunny, Quiet Room in Greenpoint    1641537   
25778  20624541      Modern apartment in the heart of Williamsburg   10132166   
25794  20639628  Spacious comfortable master bedroom with nice ...   86327101   
25795  20639792  Contemporary bedroom in brownstone with nice view   86327101   
25796  20639914       Cozy yet spacious private brownstone bedroom   86327101   
26259  20933849                              the best you can find   13709292   
26841  21291569  Coliving in Brooklyn! Modern design / Shared room  101970559   
26866  21304320             Best Coliving space ever! Shared room.  101970559   

          host_name neighbourhood_group       neighbourhood  latitude  \
23161      Kimberly            Brooklyn  Bedford-Stuyvesant  40.69023   
25433        Anisha               Bronx     East Morrisania  40.83296   
25634  Martial Loft            Brooklyn            Bushwick  40.69467   
25753        Lauren            Brooklyn          Greenpoint  40.72462   
25778       Aymeric            Brooklyn        Williamsburg  40.70838   
25794       Adeyemi            Brooklyn  Bedford-Stuyvesant  40.68173   
25795       Adeyemi            Brooklyn  Bedford-Stuyvesant  40.68279   
25796       Adeyemi            Brooklyn  Bedford-Stuyvesant  40.68258   
26259        Qiuchi           Manhattan         Murray Hill  40.75091   
26841        Sergii            Brooklyn            Bushwick  40.69211   
26866        Sergii            Brooklyn            Bushwick  40.69166   

       longitude        room_type  price  minimum_nights  number_of_reviews  \
23161  -73.95428     Private room      0               4                  1   
25433  -73.88668     Private room      0               2                 55   
25634  -73.92433     Private room      0               2                 16   
25753  -73.94072     Private room      0               2                 12   
25778  -73.94645  Entire home/apt      0               5                  3   
25794  -73.91342     Private room      0               1                 93   
25795  -73.91170     Private room      0               1                 95   
25796  -73.91284     Private room      0               1                 95   
26259  -73.97597  Entire home/apt      0               3                  0   
26841  -73.90670      Shared room      0              30                  2   
26866  -73.90928      Shared room      0              30                  5   

      last_review  reviews_per_month  calculated_host_listings_count  \
23161  2018-01-06               0.05                               4   
25433  2019-06-24               2.56                               4   
25634  2019-05-18               0.71                               5   
25753  2017-10-27               0.53                               2   
25778  2018-01-02               0.15                               1   
25794  2019-06-15               4.28                               6   
25795  2019-06-21               4.37                               6   
25796  2019-06-23               4.35                               6   
26259  2019-07-08               0.00                               1   
26841  2019-06-22               0.11                               6   
26866  2019-05-24               0.26                               6   

       availability_365  
23161                28  
25433               127  
25634                 0  
25753                 0  
25778                73  
25794               176  
25795               232  
25796               222  
26259                 0  
26841               333  
26866               139

### С типом жилья private room посчитал что после 3000 долларов просто слишком завышено и скорее нереально
### С типом жилья shared room посчитал что после 1000 долларов тоже нереально
### И Удалим жилье где прайс стоит 0


In [39]:
# Удаляем Shared room с ценой >= 1000
df.drop(df[(df['room_type'] == 'Shared room') & (df['price'] >= 1000)].index, inplace=True)

# Удаляем Private room с ценой >= 3000
df.drop(df[(df['room_type'] == 'Private room') & (df['price'] >= 3000)].index, inplace=True)

# Удаляем строки, где цена == 0
df.drop(df[df['price'] == 0].index, inplace=True)

df.reset_index(drop=True, inplace=True)

df.isnull().sum()

id                                0
name                              0
host_id                           0
host_name                         0
neighbourhood_group               0
neighbourhood                     0
latitude                          0
longitude                         0
room_type                         0
price                             0
minimum_nights                    0
number_of_reviews                 0
last_review                       0
reviews_per_month                 0
calculated_host_listings_count    0
availability_365                  0
dtype: int64

In [40]:
df.head()

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0  2019-07-08   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2               0.00                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

In [41]:
# Проверка на средную и медиану по типу жилья 

print("Количество значений по типу жилья:")
print(df["room_type"].value_counts())

print()
print('Медиана по типу жилья:')
print("Entire home/apt:", df[df['room_type'] == "Entire home/apt"]['price'].median())
print("   Private room:", df[df['room_type'] == "Private room"]['price'].median())
print("    Shared room:", df[df['room_type'] == "Shared room"]['price'].median())

print() 
print('Среднее по типу жилья:')
print("Entire home/apt:", df[df['room_type'] == "Entire home/apt"]['price'].mean())
print("   Private room:", df[df['room_type'] == "Private room"]['price'].mean())
print("    Shared room:", df[df['room_type'] == "Shared room"]['price'].mean())

Количество значений по типу жилья:
room_type
Entire home/apt    25407
Private room       22309
Shared room         1154
Name: count, dtype: int64

Медиана по типу жилья:
Entire home/apt: 160.0
   Private room: 70.0
    Shared room: 45.0

Среднее по типу жилья:
Entire home/apt: 211.81091825087574
   Private room: 87.36568201174414
    Shared room: 66.1161178509532
